### ISSUE TITLE
[RFC] How to add queue number support (specify which pin of a widget to connect )?

### ISSUE THREAD

**Issue by aiChaoSONG** [thesofproject/linux#3684]
## Background

The IPC4 SOF kernel now has basic support to the closed source firmware. Some processing components from close source firmware support multiple inputs or outputs, at the time we want to connect to the multiple inputs/outputs component (in firmware), we need to provide a number to indicate which input/output pin we want to connect, this number is called source_queue or sink_queue. An initialization IPC for that pin is also required, and also for those pins you don't connect to, you don't need to initialize them.

In current implementation, we hardcode the source_queue and sink_queue to 0: https://github.com/thesofproject/linux/blob/f7d6f01d0d16e84ee69544ea4055e7921334e0d2/sound/soc/sof/ipc4-topology.c#L1448
which limit us to those single input/output components.

To fix the issue, we need extra information from topology to indicate which pin of the source component is connect to which pin of the sink component. The SOC level route struct listed below has no extra place for us to store the source_queue and sink_queue.
```
struct snd_soc_dapm_route {
	const char *sink;
	const char *control;
	const char *source;

	/* Note: currently only supported for links where source is a supply */
	int (*connected)(struct snd_soc_dapm_widget *source,
			 struct snd_soc_dapm_widget *sink);

	struct snd_soc_dobj dobj;
};
```

***Note: Below solutions come from me and friends with limited ASOC knowledge, you can add more, and later I can make a design document from this RFC for the queue support.***

### Solution 1: Add source_queue and sink_queue everywhere

We could add and handle source_queue and sink_queue to route struct in alsa-utils, alsa-lib, asoc-topology, so that in topology we can write something like this:
```	
route."1" {
		source	"copier.host.1.1"
                source_queue    0
		sink	"copier.SSP.2.1"
                sink_queue    0
	}
```
I just want to give a mention to this solution, it will be a nightmare to upstream all the changes for this solution.

### Solution 2: Count component connection

In snd_sof_widget struct, we will introduce a new field `connection_count` to count the connection to this widget. At the first connection, the connection_count will be zero, we will use it as queue number, at the second connection, the connection_count will be one. This solution add limitation to the route order. for example: 
```
	route.1 {
		source 'mixin.4.1'    # source_queue = 0
		sink 'mixout.3.1'      # in this case, mixout.3.1 is connect to pin 0 of mixin.4.1 
	}
	route.2 {
		source 'mixin.4.1'    # source_queue = 1
		sink 'mixout.2.1'
	}
```
is different from 
```
	route.1 {
		source 'mixin.4.1'    # source_queue = 0
		sink 'mixout.2.1'      
	}
	route.2 {
		source 'mixin.4.1'    # source_queue = 1
		sink 'mixout.3.1'      # in this case, mixout.3.1 is connect to pin 1 of mixin.4.1 
	}
```
### Solution 3: Encode queue number in route->source and route->sink

With this solution, we will write route like this:
```	
route."1" {
		source	"copier.host.1.1:1"
		sink	"copier.SSP.2.1:2"        # connect pin 1 of copier.host.1.1 to pin 2 of copier.SSP.2.1
	}
```
We need extra parse of widget name from the route. I have an implementation for this solution based on the latest sof-dev
[queue_in_route.diff.txt](https://github.com/thesofproject/linux/files/8851634/queue_in_route.diff.txt), it is also tested on TGL_NOCODEC device. Compatibility is considered in my implementation, if the queue number is 0, simply write the route as before.

### Solution 4: Allocate source and sink queues dynamically, except when fixed bindings are required
For components that have no special pin(s), we will use solution 2 with increasing queue number, for example, copier, the first connection to a copier takes queue 0, and the second takes queue 1.

For components that have special pins, for example, smart_amplifier, we will add add two new tokens, input_pins/output_pins to denote how many IO pins a component has. We also add two pin arrays in topology to denote what type of component the pin can connect to.
For example, smart_amp, `input_pins = 2, output_pins = 1, input_pin_array=[copier.host copier.SSP]` (pin0 has to be connected with a copier.host)

in the kernel, we add one more function sof_route_get_widget_queues(source, sink, int *source_queue, int *sink_queue), or one more widget ops: sof_widget_get_queue(this_widget, widget_connect(ed)_to/from_this_widget, connection_dir(INPUT/OUTPUT));

This solution allow us to only modify/add code in sof layer.

### Solution 5..n: Other solution from SOF folks

@plbossart @ranj063 @RanderWang @libinyang @bardliao @kv2019i @lyakh @ujfalusi

**Comment by RanderWang** [comment]
thanks for explanation. There is no queue in IPC3 design, just one to one connection. We have two types of method to make it work: (1) assign queue id in topology. Option1 is hardest way and we don't need to do it with option2 in topology, just in driver. Option3 is the best. (2) calculate queue in driver when bind modules. use a index++ to assign queue. This is done by windows. we don't need to modify ASoC code.

**Comment by ujfalusi** [comment]
@aiChaoSONG, in ASoC DAPM the _pin_ is the widget. In other words: `copier.host.1.1.1` should be the widget and not `copier.host.1.1` (plus some index/pin number).
I guess this is solution 3, right?

**Comment by aiChaoSONG** [comment]
@ujfalusi Not quite get you point, Let's take copier as an example,  the closed source cavs copier can support up to four outputs, the queue number is used to indicate which output is used in a route. 

let's say the `queue 0` and `queue 1` are not interchangeable, `queue 0` must be connected to `copier.host.7.1`, and `queue 1` must be connected to `smart_amp.1.1`, we need a number to tell us which queue to connect, right? Solution 3 just appends the `:queue_number` to the widget name, this format is only used in route, the widget is still `copier.SSP.8.1`. The drawback is clear, at the place we want widget name, but we only have route, we need to extract widget name from the route, namely `widget_name:queue_number`.

![image](https://user-images.githubusercontent.com/52959957/172404641-4a0cc4f4-e434-4651-b782-edce78c792eb.png)

**Comment by plbossart** [comment]
@ujfalusi the DAPM concept of 'pin' is more for physical input/outputs at the hardware level, it's not intended to the best of my knowledge to represent components with multiple input/outputs.

@aiChaoSONG DAPM does not really have a concept of data transport, it's largely based on propagation of events between entities. That said, there are existing ways to control e.g. a MUX with a kcontrol, so that the user/application can specify which of the inputs get selected, I would look in this direction.

I also wonder if this really belongs in the topology btw. When a component is linked to another, can't we just allocate a queue in the driver? Why would that be set in stone in the topology, is there really any technical merit for doing this?

@lgirdwood @ranj063 this is a really important topic, please chime in.

**Comment by aiChaoSONG** [comment]
> I also wonder if this really belongs in the topology btw. When a component is linked to another, can't we just allocate a queue in the driver? Why would that be set in stone in the topology, is there really any technical merit for doing this?

@plbossart Please take a look at solution 2, in solution 2, I put some emphasis to the order of route connection. For copier, the four supported outputs are of the same equality, it's OK to allocate queue increasingly. But we have some widgets that differentiate the outputs,  For example, the closed source KPB component, the order matter there.

**Comment by kv2019i** [comment]
This will probably need discussion on alsa-devel. This capability is very generic and probably beneficial to other ASoC users.  E.g. PipeWire has a concept of "port" -> https://docs.pipewire.org/page_objects_design.html . Ack to @plbossart that DAPM is not defining data flow as such, but DAPM routes are the closest standard way to describe graphs in ASoC, so I'd continue to extend this. Otherwise we end up defining a completely parallel graph description method (for programmable DSPs).

Personally "solution 3" seems most promising, but given we have to introduce concept of "port" in ASoC, "Solution 1" might get traction as well.

**Comment by ujfalusi** [comment]
@plbossart, @kv2019i, while it is true that DAPM is not defining data flow, it is defining an audio path (Dynamic Audio Power Management). It was designed to deal with analog domain (within codecs usually).
It defines a path which needs to be enabled/disabled dynamically and on demand

Traditionally you would have `widget` for each connectable point in your system and then just connect the dots to make the graph.
Encoding an input or output port within a route is new, I guess upstream will ask why we can not have the points described and connect them.
```
                  |--> copier.SSP.8.1.0 --> copier.host.7.1.0 --> copier.host.7.1
copier.SSP.8.1 -->|
                  |--> copier.SSP.8.1.1 --> smart_amp.1.1
                  |--> copier.SSP.8.1.2 (if not used, no need to define it)
```
The IPC would be sent from `copier.SSP.8.1.1` and `copier.host.7.1.0` widget when that route is enabled.

Option 3 actually not that bad, but what the core would see might not right:
```
                  |--> copier.host.7.1
copier.SSP.8.1 -->|
                  |--> smart_amp.1.1
```
If the copier.host.7.1 path is active then DAPM _might_ not bother to power on copier.SSP.8.1 when the smart_amp.1.1 path got activated since it is already active. Are we going to send the needed IPC with to enable the copier.SSP.8.1's port 1?

Edit: s/port/queue

**Comment by plbossart** [comment]
@ujfalusi makes an excellent point. If a DAPM component is active, there's no way to intercept any DAPM event to enable additional output queues on that component when another path becomes active. Edit: same for input queues for e.g. a mixer component already processing data on a different path.

Things could also be interesting if queues are represented as separate widgets, in that output queues would be configured AFTER the component becomes active but input queues BEFORE the component becomes active - due to the graph walk and precedence.

@aiChaoSONG can you please clarify when the queue IPC configuration is supposed to be sent?

**Comment by plbossart** [comment]
Answering to my own question, the queues are supposed to be set-up during the 'bind' step

````c
static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *sroute)
{
	struct snd_sof_widget *src_widget = sroute->src_widget;
	struct snd_sof_widget *sink_widget = sroute->sink_widget;
	struct sof_ipc4_fw_module *src_fw_module = src_widget->module_info;
	struct sof_ipc4_fw_module *sink_fw_module = sink_widget->module_info;
	struct sof_ipc4_msg msg = {{ 0 }};
	u32 header, extension;
	int src_queue = 0;
	int dst_queue = 0; <<< THIS IS NOT GOOD!
	int ret;

	dev_dbg(sdev->dev, "%s: bind %s -> %s\n", __func__,
		src_widget->widget->name, sink_widget->widget->name);

	header = src_fw_module->man4_module_entry.id;
	header |= SOF_IPC4_MOD_INSTANCE(src_widget->instance_id);
	header |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_BIND);
	header |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
	header |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);

	extension = sink_fw_module->man4_module_entry.id;
	extension |= SOF_IPC4_MOD_EXT_DST_MOD_INSTANCE(sink_widget->instance_id);
	extension |= SOF_IPC4_MOD_EXT_DST_MOD_QUEUE_ID(dst_queue);
	extension |= SOF_IPC4_MOD_EXT_SRC_MOD_QUEUE_ID(src_queue);

	msg.primary = header;
	msg.extension = extension;

	ret = sof_ipc_tx_message(sdev->ipc, &msg, 0, NULL, 0);
	if (ret < 0)
		dev_err(sdev->dev, "%s: failed to bind modules %s -> %s\n",
			__func__, src_widget->widget->name, sink_widget->widget->name);

	return ret;
}

````

**Comment by ujfalusi** [comment]
@plbossart, which is in PCM path (initiated from `sof_pcm_hw_params()` if I'm looking at it right, so it is not DAPM?
In light of this, solution 3 looks like to have the least overall controversy. DAPM is ignored anyways for the path setup.

**Comment by plbossart** [comment]
@ujfalusi Solution 3 requires changes to sound/soc/soc-dapm.c which isn't going to go well upstream...

**Comment by plbossart** [comment]
Maybe a complete brain fart but the existing soc_dapm_route is represented as follows

````
/*
 * DAPM audio route definition.
 *
 * Defines an audio route originating at source via control and finishing
 * at sink.
 */
struct snd_soc_dapm_route {
	const char *sink;
	const char *control;
	const char *source;
...
}
````

@ranj063 Does anything prevent us from using the 'control' string to represent a binding between pin X of the source and pin Y of the sink? Or to create a pretend/fixed control to represent the binding?

**Comment by RanderWang** [comment]
Do we want to use static queue setting ? If a module use queue 3 defined in topology to work with a sink module and the sink is used only in some cases, should we use queue 0 or queue 3 in topology ?

**Comment by bardliao** [comment]
Can copier.SSP.8.1.0 and copier.SSP.8.1.1 be independent widget instead of part of copier.SSP.8.1? And copier.SSP.8.1 could be a mixer which can connect to copier.SSP.8.1.0 and/or copier.SSP.8.1.1. 
We can for example send creation IPC on the copier.SSP.8.1 widget and send binding IPC on the copier.SSP.8.1.0/copier.SSP.8.1.1 widgets.
We won't need to touch the dapm code if the above idea is doable.

**Comment by ujfalusi** [comment]
@bardliao, I was trying to suggest something like that. First I thought that DAPM is involved, but this should be doable with the open coded path walking in SOF.

**Comment by plbossart** [comment]
@bardliao @ujfalusi this idea of independent widgets for queues would require three routes when we have one today, and then the IPC for the bind would involve the actual sink/source - it seems a bit complicated to me also to define queue IDs in independent widgets that need to be consistent for all widgets connected to a source or sink.

**Comment by aiChaoSONG** [comment]
@ranj063 Can you comment in this RFC?

**Comment by ranj063** [comment]
> > I also wonder if this really belongs in the topology btw. When a component is linked to another, can't we just allocate a queue in the driver? Why would that be set in stone in the topology, is there really any technical merit for doing this?
> 
> @plbossart Please take a look at solution 2, in solution 2, I put some emphasis to the order of route connection. For copier, the four supported outputs are of the same equality, it's OK to allocate queue increasingly. But we have some widgets that differentiate the outputs, For example, the closed source KPB component, the order matter there.

@aiChaoSONG @plbossart I like option 2 the best. I do not think there's any merit in encoding the queues in topology at all. What did you really mean by that last sentence @aiChaoSONG ?

**Comment by aiChaoSONG** [comment]
@ranj063 I just want to give an example that option 2 adds a limitation to route definition in topology.
See the picture below,  Queue 0 should connect to PlaybackHost, and Queue 1 should connect to CaptureDaiCopier, we will write route:
```
route."1" {
    source   PlaybackHost
    sink     SmartAmp      // the route PlaybackHost->SmartAmp appear first, so it takes queue 0 of SmartAmp
}
route."2" {
    source   CaptureDaiCopier
    sink     SmartAmp     the route CaptureDaiCopier->SmartAmp appear second, so it takes queue 1 of SmartAmp
}
```
but below will define a different and wrong route, because queue 0 is for the playback input, and queue 1 is for feedback
```
route."1" {
    source   CaptureDaiCopier
    sink     SmartAmp     the route CaptureDaiCopier->SmartAmp appear second, so it takes queue 0 of SmartAmp

}
route."2" {
    source   PlaybackHost
    sink     SmartAmp      // the route PlaybackHost->SmartAmp appear first, so it takes queue 1 of SmartAmp
}
```

![image](https://user-images.githubusercontent.com/52959957/174267384-266deaa5-706e-4f11-ba09-6469c33666b2.png)

The is also one issue with option 2, take copier as an example, it supports up to 4 outputs, what if we only want use 1 and 4? I raise this question is because the comment from cavs_fw:
> Pin format arrays may be non-continuous 
> i.e. may contain pin #0 format followed by pin 2 format in case pin 1 will not be in use. 
> FW assigned format of the pin based on pin_index, not on a position of the item in the array. Applies to both input and output pins.

**Comment by plbossart** [comment]
@aiChaoSONG in a 'classic' graph system, if the pins are dedicated to a specific task they have to be strongly 'typed', with qualifiers, MIME-types, capabilities, etc, so that you don't connect pins of different nature. The codec feedback is e.g. NOT PCM, and that should be very clear.
It's not just a problem of defining the indices but exposing what each pin is used for. We will need this when we enable compressed offload.

**Comment by ranj063** [comment]
> @aiChaoSONG in a 'classic' graph system, if the pins are dedicated to a specific task they have to be strongly 'typed', with qualifiers, MIME-types, capabilities, etc, so that you don't connect pins of different nature. The codec feedback is e.g. NOT PCM, and that should be very clear. It's not just a problem of defining the indices but exposing what each pin is used for. We will need this when we enable compressed offload.

@plbossart do you mean this rules out option 2? Or are you suggesting we add pin capabilities to a widget (struct snd_sof_widget) and match them to assign the pin in the driver when binding modules?

**Comment by plbossart** [comment]
> > @aiChaoSONG in a 'classic' graph system, if the pins are dedicated to a specific task they have to be strongly 'typed', with qualifiers, MIME-types, capabilities, etc, so that you don't connect pins of different nature. The codec feedback is e.g. NOT PCM, and that should be very clear. It's not just a problem of defining the indices but exposing what each pin is used for. We will need this when we enable compressed offload.
> 
> @plbossart do you mean this rules out option 2? Or are you suggesting we add pin capabilities to a widget (struct snd_sof_widget) and match them to assign the pin in the driver when binding modules?

Option 2 (counting component connection) cannot work unless the connections are typed.

For example, in the case of a mixer, all connections are of the same type of there's no real need to pre-allocate queue numbers. We could perfectly use the next available queue when binding.
Same in the case of a demux, all queues behave in the same manner, so we could pick one as needed.

The case of the smart-amp component is 'special', some connections are more equal than others and need to be handled differently.

**Comment by ranj063** [comment]
> The case of the smart-amp component is 'special', some connections are more equal than others and need to be handled differently.

@plbossart then how about we qualify components with max output pins and connection type in the topology. So regular components like mizer/demux, this can de default type and possibly even need to additional info in topology. But for component like smart amp and the dai/host copiers, we can add the pin connection type so that we can choose which pin to connect to when binding these modules. This is a bit more work but this would allow for future usages and extensions as well

**Comment by plbossart** [comment]
This is worthy of an interactive session @ranj063. If you look at the diagram above, the capture dai copier is used, so it would need to support multiple types of data and propagate that what generated until a connection point.

FWIW, this is similar to what USB defined as 'clusters', information on number of channels and type that is propagated along the graph. The actual sampling rates and frequencies are omitted on purpose to give the implementation some slack, so it's really high level information only.

**Comment by plbossart** [comment]
I'd like to suggest

Solution 4: Allocate source and sink queues dynamically, except when fixed bindings are required

In the vast majority of cases, we can let the kernel allocate source/sink queues on a first-available basis.

Only in the case of specific components do we need to have a special binding between component queues. We would need to add this information explicitly in the topology (location TBD)

**Comment by aiChaoSONG** [comment]
@plbossart Ranjani has talked to me, and I summarized the point and some implementation details in solution 4 in the first comment box, please take a look and comment

**Comment by ranj063** [comment]
@aiChaoSONG @plbossart this is what I had in mind 
[0001-topology2-add-pin-info-to-copier-widget.txt](https://github.com/thesofproject/linux/files/8963235/0001-topology2-add-pin-info-to-copier-widget.txt)

I showed how to add the pin info for the copier widget here but in reality its only needed for special widgets for smart amp. The driver needs to parse the pin_binding_widget token array for each widget (if it exists) and save it.

**Comment by plbossart** [comment]
@ranj063 can you clarify this part of the topology, did you really mean 'input' and 'output' for valid values? Why not sink/source to be consistent.

````
# type can be either source/sink
+	DefineAttribute."type" {
+		type	"string"
+		constraints {
+			!valid_values [
+				"output"
+				"input"
+			]
+		}
````

**Comment by ranj063** [comment]
@plbossart I originally intended for it to be sink/source but changed them to input/putput to be consistent with the num_input_pins/num_output_pins. I prefer sink/source too and maybe we can change to source/sink everywhere

**Comment by plbossart** [comment]
Yes, num_sink_pins and num_source_pins are better, otherwise we'll be confused.

**Comment by ranj063** [comment]
> Yes, num_sink_pins and num_source_pins are better, otherwise we'll be confused.

makes sense. I will post the kernel patch (not tested) today as well just to show how the tokens will be used.

**Comment by ranj063** [comment]
[0001-ipc4-topology-Add-IPC-widget-ops-for-get-put-queue-I.txt](https://github.com/thesofproject/linux/files/8973016/0001-ipc4-topology-Add-IPC-widget-ops-for-get-put-queue-I.txt)

@aiChaoSONG this is what I was planning for assigning queue ID in the kernel. I added a small snipper for the smartamp as well just to show you what I meant

**Comment by ranj063** [comment]
@aiChaoSONG I thought a bit more about my kernel side implementation and I think we can in fact avoid the new ops for get/put queue I'd Nd have just 2 options. If the pinbinding_widget array for a widget is null because there's nothing specified in the topology, we use the generic dynamic quieue allocation. Otherwise, we use the mapping provided by topology.

**Comment by aiChaoSONG** [comment]
@ranj063 Some more we need consider
1. Do we want to support `Not Connected`? for example: for copier, we only connect pin 1 and pin 4, [ mixin.1.1, NC, NC, mixin.2.1 ]?  (this is supported by cavs firmware and windows)
2. Do you know if we can initialize all pins of a widget with a single MOD_INIT_INSTANCE IPC?  if no, we need extra ops to initialize other pins. 
3. For loadable module create by IADK, the ipc blob contains two parts, pt1: sof_ipc4_base_module_cfg, pt2: sof_ipc4_base_module_cfg_ext
```
struct sof_ipc4_base_module_cfg_ext
{
	uint16_t in_pin_count;
	uint16_t out_pin_count;
	uint8_t reserved[12];
	struct sof_ipc4_input_pin_format input_pins[0];
	struct sof_ipc4_output_pin_format output_pins[0];
};

struct sof_ipc4_input_pin_format
{
	uint32_t pin_index;
	uint32_t ibs;
	struct sof_ipc4_audio_format audio_fmt;
};

struct sof_ipc4_output_pin_format
{
	uint32_t pin_index;
	uint32_t obs;
	struct sof_ipc4_audio_format audio_fmt;
};
```
I am thinking to put the struct `sof_ipc4_input/output_pin_format` in topology like this, or should we duplicate the pin_format from sof_ipc4_base_module_cfg?
```
			Object.Base.pin_desc {
				output.1 {
					bind 	"gain.1.1"
                                        pin_index       0
                                        obs                  384
                                        pin_format {
                                                sampling_frequency
                                                bit_depth
                                                ch_map
                                                ch_cfg
                                                interleaving_style
                                                fmt_cfg
                                        }
				}
				input.1 {
					bind 	"copier.host.1.1"
                                         ....
				}
			}
```

**Comment by ranj063** [comment]
> @ranj063 Some more we need consider
> 
> 1. Do we want to support `Not Connected`? for example: for copier, we only connect pin 1 and pin 4, [ mixin.1.1, NC, NC, mixin.2.1 ]?  (this is supported by cavs firmware and windows)
If a copier has restriction in terms which pins should be connected, the output pins must be specified in topology in the copier instance so it only applies to that copier and not all copiers.

> 2. Do you know if we can initialize all pins of a widget with a single MOD_INIT_INSTANCE IPC?  if no, we need extra ops to initialize other pins.
No idea what you mean by initializing pins. Can you please elaborate?

> 3. For loadable module create by IADK, the ipc blob contains two parts, pt1: sof_ipc4_base_module_cfg, pt2: sof_ipc4_base_module_cfg_ext. 

It looks like we shouod just use the audio format in the base_cfg.

**Comment by bardliao** [comment]
@ranj063 @aiChaoSONG I created https://github.com/thesofproject/linux/pull/3737 based on [0001-ipc4-topology-Add-IPC-widget-ops-for-get-put-queue-I.txt](https://github.com/thesofproject/linux/files/8973016/0001-ipc4-topology-Add-IPC-widget-ops-for-get-put-queue-I.txt), we can discuss there.

**Comment by aiChaoSONG** [comment]
# Widget Queue ID Support
## Background
An IPC4 SOF/Closed-Source-cAVS widget may have one of below source/sink pin configurations:
1. One source pin, one sink pin (example: GAIN/PGA/PEAKVOL)
2. One source pin, multiple sink pins (example: COPIER/DEMUX/MIXIN)
3. Multiple source pins, one sink pin (example: MUX/MIXOUT/AEC/SMART_AMP)
4. Multiple source pins, multiple sink pins (rare, but let's not deny this configuration)

In route setup, what we have is the snd_sof_widget struct for source widget and sink widget, but for widgets with multiple source pins or sink pins, we will also need the queue ID to specify which pin we want to connect during route binding. 

## The Two Types of Pin Bindings
### Special Pin Binding
We say a widget requires special pin binding if
- it has multiple source pins or sink pins, and the pins need to be connected to specific widget, and thus cannot be allocated increasingly in route setup. For example, the SMART_AMP widget has two source pins, the feedback source pin differs from the other pin because it has to be connected to capture DAI copier to get codec feedback, the algorithm within this widget will not work properly if pins are connected wrongly. 
- it has multiple source pins or sink pins, the pins don't need to be connected to specific widget (or widget pin), but we want to leave some pin(s) between two connected pins dangling. For example, a COPIER widget supports up to 4 sink pins, we only want to connected to pin0 and pin3, pin1 and pin2 are left dangling.

### Usual Pin Binding
We say a widget has usual pin binding if
- it has only one source pin and one sink pin.
- it has multiple source pins or sink pins, but the pins can be treated equally, and we don't leave any pin dangling or dangling pins are all after connected pins, so that pins can be allocated increasingly one by one. For example, a COPIER widget supports up to 4 sink pins, and the pins can be allocated increasingly, so that at the first time it is connected, we use pin0, and at the second time, we use pin1, etc.

## The Two Queue ID Allocation Methods

There are two methods to allocate queue IDs, `the IDA method` for usual pin binding and `the binding array method` for special pin binding.
- **The IDA method**: for usual pin binding, we will use kernel ID Allocation API to allocate widget queue ID increasingly during route setup, we don't need to do anything in topology.
- **The binding array method**: for special pin binding, we will parse two pin binding arrays (src_pin_binding / sink_pin_binding) for the widget from topology:

    - **src_pin_binding**: the pin binding array for source pins of a widget, it takes the form of `[widget_name_connected_to_pin0, widget_name_connected_to_pin1, ...]`

    - **sink_pin_binding**: the pin binding array for sink pins of a widget, it also takes the form of `[widget_name_connected_to_pin0, widget_name_connected_to_pin1, ...]`
We will need to define the pin binding array in topology.

All source pins or sink pins of a widget has to use the same pin binding method. You should not allocate queue ID with special pin binding method for one source pin, and uses IDA for other source pins, this rule applies to sink pins, too. But you may allocate source pin with the binding array method and allocate sink pin with the IDA method.

## Implementation

### Topology

Two tokens `num_source_pins` and `num_sink_pins` are added, so that we know how many source pins and sink pins a widget can support. They should be defined per widget. Both pin binding method rely on these two values.

#### Special Pin Binding Case

Due to the fact that the element of the src_pin_binding/sink_pin_binding has to be stored in topology as 

```c
struct snd_soc_tplg_vendor_string_elem {
 __le32 token;
 char string[SNDRV_CTL_ELEM_ID_NAME_MAXLEN];
}
```

we will need two extra tokens, let us say: `src_pin_binding_wname` and `sink_pin_binding_wname` , to parse the widget names to the binding arrays. 
For completeness, we provide two pin binding classes:

```
Class.Base."src_pin_binding" {
	DefineAttribute."instance" {}

	DefineAttribute."src_pin_binding_wname" {
		token_ref "sof_tkn_comp.string"
	}
        ...
}

Class.Base."sink_pin_binding" {
	DefineAttribute."instance" {}

	DefineAttribute."sink_pin_binding_wname" {
		token_ref "sof_tkn_comp.string"
	}
        ...
}
```
So that we can add pin binding array like this:

```
Object.Widget.copier.1 {
	stream_name 'SSP0 Playback'
	num_sink_pins		4
	Object.Base.sink_pin_binding.1 {
		sink_pin_binding_wname	"NC"
	}
	Object.Base.sink_pin_binding.2 {
		sink_pin_binding_wname	"copier.SSP.2.1"
	}
	Object.Base.sink_pin_binding.3 {
		sink_pin_binding_wname	"NC"
	}
	Object.Base.sink_pin_binding.4 {
		sink_pin_binding_wname	"NC"
	}
}
```

Note that once you use the pin binding method for one source pin, you should also define the pin binding for all other source pins, if a pin is not used, just give the value "NC" (means 'not connected').

#### Usual Pin Binding Case
If you don't want to use special pin binding, don't define pin binding in topology, the kernel will allocate queue ID with IDA dynamically.

### Kernel
- `num_source_pins` and `num_sink_pins` are added to snd_sof_widget structure, so that we know how many pin binding to parse
- src_pin_binding` and `sink_pin_binding` are added to snd_sof_widget for the pin binding array
- src_queue_ida and sink_queue_ida are added to snd_sof_widget for the ida allocation method.

`num_source_pins`, `num_sink_pins`, `src_pin_binding` and `sink_pin_binding` are parsed from topology in `widget_ready` callback.

After parsing above tokens from topology, we have all the information for widget binding, so that in route setup, we will first check pin binding array, and then
- pin binding not empty: iterate the array to get the correct queue ID, if pin binding is defined but queue ID cannot find, fail because we don't want to mixed use both methods.
- pin binding empty: use IDA to allocate the queue ID.

## Pin Binding Examples
### SMART_AMP
Route graph for SMART_AMP is below:
![image](https://user-images.githubusercontent.com/52959957/180944485-e1fa27d0-696f-4f2e-b008-010ac78e2aea.png)
```
Object.Widget.smart_amp.1 {
        # for queue 0
	Object.Base.src_pin_binding.1 {
		src_pin_binding_wname	"copier.host.1.1"
	}
        # for queue 1
	Object.Base.src_pin_binding.2 {
		src_pin_binding_wname	"copier.SSP.8.1"
	}
}
```
The num_source_pins = 2 and num_sink_pins = 1  are defined in the SMART_AMP class.

### MIXOUT Example 1
Route graph for MIXOUT is below:
![image](https://user-images.githubusercontent.com/52959957/180946022-d883c68a-76d4-4b4d-bbc0-b09673d51836.png)
The num_source_pins = 8 and num_sink_pins = 1  are defined in the MIXOUT class.
MIXOUT widget supported 8 sources and 1 sink, here we only used source pin 0 and source pin 1. Obviously there is no special pin binding and no pins are dangling, so we don't need to define pin binding in topology, the kernel will help to allocate dynamically with IDA.

### MIXOUT Example 2
Route graph for MIXOUT is below:
![image](https://user-images.githubusercontent.com/52959957/180946709-cdbb53ee-ff13-4d8e-bd97-fb959efa6df5.png)
```
Object.Widget.mixout.1 {
        # for queue 0
	Object.Base.src_pin_binding.1 {
		src_pin_binding_wname    "copier.host.1.1"
	}
	Object.Base.src_pin_binding.2 {
		src_pin_binding_wname    "NC"
	}
	Object.Base.src_pin_binding.3 {
		src_pin_binding_wname    "NC"
	}
	Object.Base.src_pin_binding.4 {
		src_pin_binding_wname    "NC"
	}
        # for queue 4
	Object.Base.src_pin_binding.5 {
		src_pin_binding_wname    "copier.host.3.1"
	}
	Object.Base.src_pin_binding.6 {
		src_pin_binding_wname    "NC"
	}
	Object.Base.src_pin_binding.7 {
		src_pin_binding_wname    "NC"
	}
	Object.Base.src_pin_binding.8 {
		src_pin_binding_wname    "NC"
	}
}
```
The num_source_pins = 8 and num_sink_pins = 1  are defined in the MIXOUT class.
MIXOUT widget supported 8 sources (0~7) and 1 sink, here we only used source pin 0 and source pin 4. Pin 1,2,3,5,6,7 are dangling, special pin binding is required here.

## Best Practice
- if you define special pin binding for a source(sink) pin of a widget, you **should** define special pin binding for all source(sink) pins, although they are not used.
- if num_source_pins or num_sink_pins is one, you **may not** define any special pin binding in topology, the kernel will help to allocate queue ID = 0 here with IDA.
- if num_souce_pins or num_sink_pins is greater than one, but special pin binding is not required, you **may not** define any special pin binding in topology, the kernel will help to allocate queue ID dynamically with IDA.
- if special pin binding is defined for a not used pin, you **may** give the value `NC` (not connected) to `src_pin_binding_wname`  or `sink_pin_binding_wname`.

## FAQ
**Q: Why pin binding instance starts with one, can we use a zero-base instance?**
The instance is a topology 2 concept, it is used mostly to distinguish widgets within the same alsaconf node, and will not be compiled to the binary. (the one that will be compiled to topology binary is the index attribute). Another usage is that is will be used to construct widget name, so that we can refer the widget in other part of the topology. 
Due to a limitation in alsatplg, we have to use one-based instance for all widget for topology 2. All other widgets now use one-based instance, let's follow this.

## Pull Requests
- The Kernel PR: https://github.com/thesofproject/linux/pull/3757
- The SOF topology PR: https://github.com/thesofproject/sof/pull/6000

**Comment by plbossart** [comment]
> In route binding, the information we have is souce widget and sink widget, their queue IDs are unknown. To acquire source and sink widget queue IDs, we add two string array in the snd_sof_widget structure, which will be parsed from topology:
> 
> * src_pin_binding: the pin binding array for source pins of a widget, it takes the form of `[widget_name_connected_to_pin0, widget_name_connected_to_pin1, ...]`
> * sink_pin_binding: the pin binding array for sink pins of a widget, it also takes the form of `[widget_name_connected_to_pin0, widget_name_connected_to_pin1, ...]`
> 
> so that in route binding, we can search the pin binding array for source widget and sink widget to get the source widget queue ID and sink widget queue ID.

so what happens if you have e.g. 4 pins and only pin0 and pin3 are connected. Does this mean that pin1 and pin2 need to be represented as not connected?

Also what happens for e.g. a mixer, are we required to assign pins manually in this way or can we let the kernel use pins dynamically?

And last, what happens in the case of the smartamp where one pin is dedicated to IV feedback and the other isn't? Do we need to describe all pins and all bindings?

You really need to describe your solution using SHALL and MAY terms, that what we can understand what is required and what is permitted.



> Object.Widget.copier.1 {
> 	stream_name 'SSP0 Playback'
> 	num_sink_pins		4
> 	Object.Base.sink_pin_binding.1 {
> 		sink_pin_binding_wname	"copier.SSP.2.1"
> 	}
> 	Object.Base.sink_pin_binding.2 {
> 		sink_pin_binding_wname	"NC"
> 	}
> 	Object.Base.sink_pin_binding.3 {
> 		sink_pin_binding_wname	"NC"
> 	}
> 	Object.Base.sink_pin_binding.4 {
> 		sink_pin_binding_wname	"NC"

Can we use a zero-base pin binding? Why would we start with pin_binding.1

Too many details missing IMHO @aiChaoSONG

**Comment by aiChaoSONG** [comment]
@plbossart I now updated the whole documentation for the queue support, hope it helps everyone to understand the issue and how we fix it, of course helps the reviewer to better review the code. still in the previous comment: 
https://github.com/thesofproject/linux/issues/3684#issuecomment-1193614746

**Comment by RanderWang** [comment]
MIXOUT should cooperate with MIXIN and it can't be connected to copier directly

**Comment by plbossart** [comment]
closing, this is implemented in PR #3757
